### PR TITLE
Offline rendering feature for browser mode (Tone v14 only) + some tests for browser mode

### PR DESCRIPTION
**Pull request by qchenevier**
This has been pretty hard to implement due to the fact that Tone.js seems to be designed to render 1 buffer at a time, not several at the same time like it is in this PR.

### For Tone v14
Tone.js uses global variables to save/modify/restore webaudio context during offline rendering. So if you render several clips at the same time, the function calls are modifying global variables concurrently. This messes up the Tone.js webaudio context so I had to implement a custom context saving/restoring.

I guess I'll contribute an issue in the Tone.js repo for that.

### For Tone v13
I couldn't get it working for several renders at the same time. In the end, I disabled the feature with Tone v13 a and put a warning in the console, saying that offline rendering in scribbletune is not available for this Tone.js version.

### Tests
I added some tests for a function which has been designed well before that. So they are unrelated to the new feature implemented (which I don't know how to test properly for now). I could have filed another PR for that, but I've put everything in the same PR out of pure laziness (that's bad I know).

### CODE REVIEW

**Comment by qchenevier** [comment]
I noticed in the end that this feature doesn't work with a clip starting with a silence, like: `{ pattern: '---x' }`. This might be related with Tone.js offline rendering function. I'll create an issue for that.

File: src/browserClip.ts [hunk 1]
```diff
@@ -111,13 +111,11 @@ const recursivelyApplyPatternToDurations = (
...
 
-/**
- * @param  {Object}
- * @return {Tone.js Sequence Object}
- * Take a object literal that may have a Tone.js player OR instrument
- * or simply a sample or synth with a pattern and return a Tone.js sequence
- */
-module.exports = (params: ClipParams) => {
+const isToneLegacy = () => {
+  return Tone.version.split('.')[0] === '13';
```
Comment (walmik) [inline src/browserClip.ts:115]: Since this is going for legacy we could even go for `return +Tone.version.split('.')[0] < 14;`

File: src/clip.ts [hunk 2]
```diff
@@ -2,7 +2,8 @@ import { isNote, shuffle, expandStr } from './utils';
 import { getChord } from './chord';
 
 /* tslint:disable:no-var-requires */
-const browserClip = typeof window !== 'undefined' && require('./browserClip');
+const browserClip =
+  typeof window !== 'undefined' && require('./browserClip').browserClip;
 
 /**
  * Get defauly params for a clip, such as root note, pattern etc
```
Comment (walmik) [inline src/clip.ts:9]: This is not part of the PR but just noticed it, since there is another comment on this PR, would you kindly fix this typo "defauly" as part of this PR.

File: dist/script_offline.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  const session = new scribble.Session();
+
+  session.createChannel({
+    instrument: 'PolySynth',
+    clips: [
+      { pattern: '[xx]', notes: 'C3 Cm-3' },
+      { pattern: '[x-]', notes: 'C3 Cm-3' },
+      { pattern: 'xx', notes: 'C3 Cm-3' },
+    ],
+    offlineRendering: true,
```
Comment (walmik) [inline dist/script_offline.js:11]: Should we add `offlineRendering` in here as well: https://github.com/scribbletune/scribbletune/blob/master/src/typings.d.ts#L156-L159

File: src/browserClip.ts [hunk 4]
```diff
@@ -248,3 +248,62 @@ module.exports = (params: ClipParams) => {
     );
   }
 };
+
+const totalPatternDuration = (pattern: any) => {
+  return recursivelyApplyPatternToDurations(
+    expandStr(pattern),
+    Tone.Ticks('4n').toSeconds()
```
Comment (walmik) [inline src/browserClip.ts:255]: Since you had add the improvement to infer correct note duration in an earlier commit, do you think we should set "4n" as a backup in case no subdiv or dur is provided by the user, for eg https://github.com/scribbletune/scribbletune/blob/master/src/clip.ts#L174

File: dist/script_offline.js (hunk 3 above)
Comment (walmik) [reply dist/script_offline.js:11]: May not be needed on second thoughts.

File: src/browserClip.ts [hunk 5]
```diff
@@ -111,13 +111,11 @@ const recursivelyApplyPatternToDurations = (
   return durations;
 };
 
-/**
- * @param  {Object}
- * @return {Tone.js Sequence Object}
```
Comment (walmik) [inline src/browserClip.ts:116]: We can retain this JS Doc style comment coz it s still applicable to `generateSequence`

File: src/typings.d.ts [hunk 6]
```diff
@@ -151,6 +151,13 @@ interface ClipParams {
    *  - See:      https://tonejs.github.io/docs/13.8.25/Volume
    */
   volume?: number;
+  /**
+   * Boolean parameter to trigger offline rendering. Use this to help
+   * If true, `scribbletune.clip` returns a `Tone.Player` with a buffer containing a pre-rendered sound of the sequence
+   * If alse, it returns a `Tone.Sequence` which does live rendering.
```
Comment (walmik) [inline src/typings.d.ts:157]: Minor typo `alse`

File: src/typings.d.ts [hunk 7]
```diff
@@ -151,6 +151,13 @@ interface ClipParams {
    *  - See:      https://tonejs.github.io/docs/13.8.25/Volume
    */
   volume?: number;
+  /**
+   * Boolean parameter to trigger offline rendering. Use this to help
```
Comment (walmik) [inline src/typings.d.ts:155]: sorry for the nit but is the sentence incomplete accidentally?

File: src/typings.d.ts [hunk 8]
```diff
@@ -151,6 +151,13 @@ interface ClipParams {
...
+   * If true, `scribbletune.clip` returns a `Tone.Player` with a buffer containing a pre-rendered sound of the sequence
+   * If alse, it returns a `Tone.Sequence` which does live rendering.
+   * Note: The rendering needs some time to complete & be saved in the buffer, before being able to playing.
```
Comment (walmik) [inline src/typings.d.ts:158]: I m a bit curious about this aspect. I noted some discussion around this on Tone's issues section too. For current PR we are all good, but maybe we should consider some way to "announce" that the rendering is done.

We have a similar issue for the sample and samples props. In a presentation I gave, I actually kept the network tab open and waited for all the samples in my app to load completely before proceeding withe demo :P

**Comment by walmik** [comment]
Great stuff @qchenevier ! Just left few minor comments, but overall good to merge. I ll create a minor version bump PR once @mickmister s PR is merged in as well.

File: src/typings.d.ts (hunk 7 above)
Comment (qchenevier) [reply src/typings.d.ts:155]: Sure. 😅 Thanks for seeing that.

File: src/typings.d.ts (hunk 8 above)
Comment (qchenevier) [reply src/typings.d.ts:158]: Yeah, I've had this concern too. I've thought about something like: allowing the user to pass a callback, which would be triggered at the end of the rendering. We may have 2 options to trigger this callback: after rendering of each clip or after rendering all the clips of the session/channel.

I didn't know what to chose for now (maybe we need to allow both ?), so I've simply put a `console.log` message at each beginning/end of rendering.

File: src/browserClip.ts (hunk 4 above)
Comment (qchenevier) [reply src/browserClip.ts:255]: Hey actually I missed that ! I should use here `subdiv` provided in params instead of hard-coding this subdivision value.

Comment (qchenevier) [reply src/browserClip.ts]: Solved in https://github.com/scribbletune/scribbletune/pull/113/commits/e5a98813cc7d71cbe4d9d1af10cfb163452340f4

File: src/typings.d.ts (hunk 7 above)
Comment (qchenevier) [reply src/typings.d.ts]: Solved in https://github.com/scribbletune/scribbletune/pull/113/commits/1ff91750cb3dad5b79bff216af47fcc83b9f781c

File: src/typings.d.ts (hunk 6 above)
Comment (qchenevier) [reply src/typings.d.ts]: Solved in https://github.com/scribbletune/scribbletune/pull/113/commits/1ff91750cb3dad5b79bff216af47fcc83b9f781c

File: src/clip.ts (hunk 2 above)
Comment (qchenevier) [reply src/clip.ts]: Solved in https://github.com/scribbletune/scribbletune/pull/113/commits/1ff91750cb3dad5b79bff216af47fcc83b9f781c

File: src/browserClip.ts (hunk 1 above)
Comment (qchenevier) [reply src/browserClip.ts:115]: That's right. I'll stick with `isToneV13` because I don't know the behavior for earlier versions of Tone.js.

Comment (qchenevier) [reply src/browserClip.ts:115]: Solved in https://github.com/scribbletune/scribbletune/pull/113/commits/1ff91750cb3dad5b79bff216af47fcc83b9f781c

File: src/typings.d.ts (hunk 8 above)
Comment (qchenevier) [reply src/typings.d.ts:158]: I'll file an issue/feature request to keep track of this need for improvement.

**Comment by qchenevier** [comment]
I did a rebase on the master branch, squashing all commits of this PR into 1, then did a `npm run build`.
